compiler/rustc_target/src/spec/mod.rs

```diff
@@ -2421,10 +2421,77 @@ impl ToJson for Target {
 }
 
 /// Either a target triple string or a path to a JSON file.
-#[derive(PartialEq, Clone, Debug, Hash, Encodable, Decodable)]
+#[derive(Clone, Debug)]
 pub enum TargetTriple {
     TargetTriple(String),
-    TargetPath(PathBuf),
+    TargetJson {
+        /// Warning: This field may only be used by rustdoc. Using it anywhere else will lead to
+        /// inconsistencies as it is discarded during serialization.
+        path_for_rustdoc: PathBuf,
+        triple: String,
+        contents: String,
+    },
+}
+
+// Use a manual implementation to ignore the path field
+impl PartialEq for TargetTriple {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::TargetTriple(l0), Self::TargetTriple(r0)) => l0 == r0,
+            (
+                Self::TargetJson { path_for_rustdoc: _, triple: l_triple, contents: l_contents },
+                Self::TargetJson { path_for_rustdoc: _, triple: r_triple, contents: r_contents },
+            ) => l_triple == r_triple && l_contents == r_contents,
```

Comment:
Minor style nit, consider:

```suggestion
            (l@Self::TargetJson { .. }, r@Self::TargetJson { .. }) => {
                l.triple == r.triple && l.contents == r.contents
            }
```

or, if the intent was to have this an exhaustive match, maybe something like this (though there isn’t much value in this…):


```suggestion
            (Self::TargetJson { path_for_rustdoc: _, triple, contents }, r@Self::TargetJson { .. }) => {
                triple == r.triple && contents == r.contents
            }
```
```

Reply:
> or, if the intent was to have this an exhaustive match

Indeed

> maybe something like this

Rustfmt puts that suggestion on four lines too.